File: python/y2019/day04.py
from aoc.core import solve, allValid, AocInput
from collections import Counter
from itertools import pairwise


def sixDigits(password: str) -> bool:
    return len(password) == 6


def increasing(password: str) -> bool:
    return not any(a > b for a, b in pairwise(password))


def potentialDouble(password: str) -> bool:
    return any(filter(lambda n: n >= 2, Counter(password).values()))


def exactDouble(password: str) -> bool:
    return any(filter(lambda n: n == 2, Counter(password).values()))


def part1(input: AocInput) -> int:
    lower, upper = input.text.split("-", 2)
    return sum(allValid(str(pw), sixDigits, increasing, potentialDouble) for pw in range(int(lower), int(upper)))


def part2(input: AocInput) -> int:
    lower, upper = input.text.split("-", 2)
    return sum(allValid(str(pw), sixDigits, increasing, exactDouble) for pw in range(int(lower), int(upper)))


def main() -> None:
    YEAR = "y2019"
    DAY = "Day04"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1, skipTest=True)
    solve(YEAR, DAY, part2, skipTest=True)

File: dotnet/2019/AdventOfCode.Y2019/Day04.cs
using AdventOfCode.Lib;

namespace AdventOfCode.Y2019;

[AocPuzzle(2019, 4, "Secure Container")]
public sealed class Day04 : IAocDay<int>
{
    public static int Part1(AocInput input)
    {
        var parts = input.Text.Split('-', 2);
        var (lower, upper) = (int.Parse(parts[0]), int.Parse(parts[1]));
        return Enumerable
            .Range(lower, upper - lower + 1)
            .Count(pw => AllValid($"{pw}", SixDigits, Increasing, PotentialDouble));
    }

    public static int Part2(AocInput input)
    {
        var parts = input.Text.Split('-', 2);
        var (lower, upper) = (int.Parse(parts[0]), int.Parse(parts[1]));
        return Enumerable
            .Range(lower, upper - lower + 1)
            .Count(pw => AllValid($"{pw}", SixDigits, Increasing, ExactDouble));
    }

    private static bool SixDigits(string password) =>
        password.Length is 6;

    private static bool Increasing(string password) =>
        password.Zip(password[1..]).Any(x => x.First > x.Second) is false;

    private static bool PotentialDouble(string password) =>
        password.GroupBy(g => g).ToDictionary(k => k.Key, v => v.Count()).Values.Any(c => c >= 2);

    private static bool ExactDouble(string password) =>
        password.GroupBy(g => g).ToDictionary(k => k.Key, v => v.Count()).Values.Any(c => c is 2);

    private static bool AllValid(string password, params Func<string, bool>[] funcs) =>
        funcs.All(f => f(password));
}